oops/backplane/__init__.py

```diff
@@ -84,204 +141,266 @@ def __init__(self, obs, meshgrid=None, time=None, inventory=None,
     @property
     def dlos_duv(self):
-        if self._filled_dlos_duv is None:
-            self._filled_dlos_duv = self.meshgrid.dlos_duv(self.time)
+        if not hasattr(self, '_dlos_duv'):
+            self._dlos_duv = self.meshgrid.dlos_duv(self.time)
 
-        return self._filled_dlos_duv
+        return self._dlos_duv
 
-    #===========================================================================
     @property
     def duv_dlos(self):
-        if self._filled_duv_dlos is None:
-            self._filled_duv_dlos = self.meshgrid.duv_dlos(self.time)
+        if not hasattr(self, '_duv_dlos'):
+            self._duv_dlos = self.meshgrid.duv_dlos(self.time)
+
+        return self._duv_dlos
 
-        return self._filled_duv_dlos
+    @property
+    def center_dlos_duv(self):
+        if not hasattr(self, '_center_dlos_duv'):
+            self._center_dlos_duv = self.meshgrid.center_dlos_duv(self.time)
+
+        return self._center_dlos_duv
+
+    @property
+    def center_duv_dlos(self):
+        if not hasattr(self, '_center_duv_dlos'):
+            self._center_duv_dlos = self.meshgrid.center_duv_dlos(self.time)
+
+        return self._center_duv_dlos
 
     ############################################################################
-    # Event manipulations
-    #
-    # Event keys are tuples (body_id, body_id, ...) where each body_id is a
-    # a step along the path of a photon. The path always begins at the Sun and
-    # and ends at the observer.
+    # Dictionary keys
     ############################################################################
 
-    #===========================================================================
+    @functools.lru_cache(maxsize=300)
     def standardize_event_key(self, event_key):
         """Repair an event key to make it suitable for indexing a dictionary.
 
-        Strings are converted to uppercase. A string gets turned into a tuple.
-        Solar illumination is added at the front if no lighting is specified.
+        The photons originate from the Sun unless otherwise indicated. An empty
+        event_key is returned as an empty tuple.
         """
 
-        # Convert to tuple if necessary
+        if not event_key:
+            return ()
+
+        # Handle an individual string
         if isinstance(event_key, str):
-            event_key = (event_key.upper(),)
+            return ('SUN<', event_key.upper())
 
-        elif type(event_key) == tuple:
-            items = []
-            for item in event_key:
-                if isinstance(item, str):
-                    items.append(item.upper())
-                else:
-                    items.append(item)
+        # Handle a tuple of strings
+        new_key = [k.upper() for k in event_key]
 
-            event_key = tuple(items)
+        # Begin with the Sun if there is no illumination already
+        if new_key[0][-1] not in ('<', '>', '-'):
```

Comment:
It doesn't replace the first item; it puts "SUN<" in front. The question is how to interpret the tuple ('MIMAS', 'SATURN:RING'). Right now, it would become ('SUN<', 'MIMAS', 'SATURN:RING'), which describes Mimas shadowing the rings. Your alternative would be to turn it into ('MIMAS>', 'SATURN:RING'), which would describe Mimas being occulted by the rings. You're right, that's also a plausible event key. However, I strongly prefer the former interpretation. My philosophy on event keys has been "don't worry about the lighting; we'll take care of it if you forgot." It follows from that, however, that if you want occultation lighting, you will always have to specify that explicitly. OK?